--- graphite/lib/carbon/client.py.orig	2014-12-31 16:49:45.000000000 +0000
+++ graphite/lib/carbon/client.py	2015-07-07 15:07:50.844870514 +0000
@@ -1,3 +1,5 @@
+# THIS FILE PATCHED BY BCPC
+
 from twisted.application.service import Service
 from twisted.internet import reactor
 from twisted.internet.defer import Deferred, DeferredList
@@ -76,7 +78,8 @@
       queueSize = self.factory.queueSize
       if (self.factory.queueFull.called and
           queueSize < SEND_QUEUE_LOW_WATERMARK):
-        self.factory.queueHasSpace.callback(queueSize)
+        if not self.factory.queueHasSpace.called:
+          self.factory.queueHasSpace.callback(queueSize)

   def __str__(self):
     return 'CarbonClientProtocol(%s:%d:%s)' % (self.factory.destination)
