--- zabbix/alertscripts/pagerduty.py.orig	2015-10-12 21:16:14.163817910 +0000
+++ zabbix/alertscripts/pagerduty.py	2015-10-12 21:16:14.163817910 +0000
@@ -72,10 +72,15 @@
     """

     EVENTS_API_BASE = "https://events.pagerduty.com/generic/2010-04-15/create_event.json"
-    proxy = "YOUR_PROXY_GOES_HERE"
+    config = json.loads(open("/etc/pagerduty.conf", "r").read())
+    if config['proxy_server_url']:
+        proxies = {"http":config['proxy_server_url'], "https":config['proxy_server_url']}
+    else:
+        proxies = {}

-    def __init__(self, api_base=EVENTS_API_BASE):
+    def __init__(self, api_base=EVENTS_API_BASE, proxies=proxies):
         self.api_base = api_base
+        self.proxies = proxies

     def submit_event(self, file_path):
         json_event = None
@@ -86,8 +91,7 @@
         retry = False

         try:
-            proxies = {"http":"http://%s" % proxy}
-            proxy_support = urllib2.ProxyHandler(proxies)
+            proxy_support = urllib2.ProxyHandler(self.proxies)
             opener = urllib2.build_opener(proxy_support, urllib2.HTTPHandler(debuglevel=1))
             urllib2.install_opener(opener)
             request = urllib2.Request(self.api_base)
@@ -247,4 +251,4 @@
     pagerduty_queue = PagerDutyQueue()
     if len(sys.argv) == 4:
         pagerduty_queue.enqueue(Zabbix(sys.argv).event())
-    pagerduty_queue.lock_and_flush_queue()
\ No newline at end of file
+    pagerduty_queue.lock_and_flush_queue()
