diff --git a/nova/compute/manager.py b/nova/compute/manager.py
index 5d86698..7f77bc9 100644
--- a/nova/compute/manager.py
+++ b/nova/compute/manager.py
@@ -420,6 +420,11 @@ def object_compat(function):
     def decorated_function(self, context, *args, **kwargs):
         def _load_instance(instance_or_dict):
             if isinstance(instance_or_dict, dict):
+                # try to get metadata and system_metadata for most cases but
+                # only attempt to load those if the db instance already has
+                # those fields joined
+                metas = [meta for meta in ('metadata', 'system_metadata')
+                         if meta in instance_or_dict]
                 instance = objects.Instance._from_db_object(
                     context, objects.Instance(), instance_or_dict,
                     expected_attrs=metas)
@@ -427,7 +432,6 @@ def object_compat(function):
                 return instance
             return instance_or_dict
 
-        metas = ['metadata', 'system_metadata']
         try:
             kwargs['instance'] = _load_instance(kwargs['instance'])
         except KeyError:
