diff --git a/nova/network/manager.py b/nova/network/manager.py
index 3e8e8b1..832fd1b 100644
--- a/nova/network/manager.py
+++ b/nova/network/manager.py
@@ -1351,13 +1351,15 @@ class NetworkManager(manager.Manager):
                 else:
                     net.gateway = current
                     current += 1
-                if not dhcp_server:
-                    dhcp_server = net.gateway
+                if dhcp_server:
+                    subnet_dhcp_server = dhcp_server
+                else:
+                    subnet_dhcp_server = net.gateway
                 net.dhcp_start = current
                 current += 1
-                if str(net.dhcp_start) == dhcp_server:
+                if str(net.dhcp_start) == subnet_dhcp_server:
                     net.dhcp_start = current
-                net.dhcp_server = dhcp_server
+                net.dhcp_server = subnet_dhcp_server
                 extra_reserved.append(str(net.dhcp_server))
                 extra_reserved.append(str(net.gateway))

