diff --git a/openstack_dashboard/dashboards/project/access_and_security/api_access/views.py b/openstack_dashboard/dashboards/project/access_and_security/api_access/views.py
index d71bb1a..b2484a8 100644
--- a/openstack_dashboard/dashboards/project/access_and_security/api_access/views.py
+++ b/openstack_dashboard/dashboards/project/access_and_security/api_access/views.py
@@ -19,6 +19,7 @@ import zipfile
 
 from django import http
 from django import shortcuts
+from django.conf import settings
 from django.template.loader import render_to_string
 from django.utils.translation import ugettext_lazy as _
 
@@ -70,6 +71,7 @@ def _get_openrc_credentials(request):
     credentials = dict(tenant_id=request.user.tenant_id,
                        tenant_name=request.user.tenant_name,
                        auth_url=keystone_url,
+                       versions=settings.OPENSTACK_API_VERSIONS,
                        user=request.user,
                        region=getattr(request.user, 'services_region') or "")
     return credentials
