--- keystone/common/ldap/core.py.bak	2015-05-12 11:44:02.589167178 -0400
+++ keystone/common/ldap/core.py	2015-05-12 11:44:09.752800114 -0400
@@ -1671,7 +1671,7 @@
                       'entries': not_deleted_nodes[:3],
                       'dots': '...' if len(not_deleted_nodes) > 3 else ''})

-    def filter_query(self, hints, query=None):
+    def filter_query(self, hints, query=''):
         """Applies filtering to a query.

         :param hints: contains the list of filters, which may be None,
