--- a/nova/network/linux_net.py   2016-03-14 16:17:32.517329066 +0000
+++ b/nova/network/linux_net.py   2016-03-15 15:48:39.712397155 +0000
@@ -15,6 +15,8 @@
 #    License for the specific language governing permissions and limitations
 #    under the License.
 
+# THIS FILE PATCHED BY BCPC
+
 """Implements vlans, bridges, and iptables rules using linux utilities."""
 
 import calendar
@@ -1127,6 +1129,7 @@
            '--dhcp-optsfile=%s' % _dhcp_file(dev, 'opts'),
            '--listen-address=%s' % network_ref['dhcp_server'],
            '--except-interface=lo',
+           '--interface=%s' % dev,
            '--dhcp-range=set:%s,%s,static,%s,%ss' %
                          (network_ref['label'],
                           network_ref['dhcp_start'],
@@ -1220,7 +1223,8 @@
     # NOTE(cfb): dnsmasq on linux only supports 64 characters in the hostname
     #            field (LP #1238910). Since the . counts as a character we need
     #            to truncate the hostname to only 63 characters.
-    hostname = fixedip.instance.hostname
+    #hostname = fixedip.instance.hostname # EDITED BY BCPC
+    hostname = 'ip-%s' % str(fixedip.address).replace('.', '-')
     if len(hostname) > 63:
         LOG.warning(_LW('hostname %s too long, truncating.') % (hostname))
         hostname = fixedip.instance.hostname[:2] + '-' +\
