--- nova/virt/libvirt/config.py.bak	2015-04-28 14:27:38.114711892 -0400
+++ nova/virt/libvirt/config.py	2015-04-28 15:31:24.791723881 -0400
@@ -1827,11 +1827,13 @@
         if self.numatune is not None:
             root.append(self.numatune.format_dom())
         if self.cpuset is not None:
-            vcpu = self._text_node("vcpu", self.vcpus)
-            vcpu.set("cpuset", hardware.format_cpu_spec(self.cpuset))
-            root.append(vcpu)
-        else:
             root.append(self._text_node("vcpu", self.vcpus))
+#        if self.cpuset is not None:
+#            vcpu = self._text_node("vcpu", self.vcpus)
+#            vcpu.set("cpuset", hardware.format_cpu_spec(self.cpuset))
+#            root.append(vcpu)
+#        else:
+#            root.append(self._text_node("vcpu", self.vcpus))

         if len(self.metadata) > 0:
             metadata = etree.Element("metadata")